examples/hello_world/client/android/java/com/google/oak/hello_world/MainActivity.java

```diff
@@ -22,59 +22,84 @@
+
     Button helloButton = findViewById(R.id.helloButton);
-    helloButton.setOnClickListener(v -> onClick());
+    helloButton.setOnClickListener(v -> helloButtonOnClick());
 
     // Set default address.
     // Android emulator forwards `10.0.2.2` to the host machine.
     // https://developer.android.com/studio/run/emulator-networking
     EditText addressInput = findViewById(R.id.addressInput);
-    addressInput.setText("10.0.2.2:8888");
-  }
-
-  @Override
-  public void onDestroy() {
-    super.onDestroy();
+    addressInput.setText("10.0.2.2:8080");
   }
 
-  public void onClick() {
+  /** Handles click events from the helloButton. */
+  public void helloButtonOnClick() {
     EditText addressInput = findViewById(R.id.addressInput);
     String address = addressInput.getText().toString();
-
-    if (!address.equals(rpcAddress)) {
-      Log.v("Oak", "Create channel to: " + address);
-      createChannel(address);
-      rpcAddress = address;
-    }
-
     Log.v("Oak", "Say Hello");
     TextView helloTextView = findViewById(R.id.helloTextView);
-    String responce = sayHello("World");
-    Log.v("Oak", "Responce is: " + responce);
-    helloTextView.setText(responce);
+
+    // TODO(#988): Move blocking call ouf of UI thread.
+    String reply = sayHello(address, "World");
+    Log.v("Oak", "Response is: " + reply);
+    helloTextView.setText(reply);
   }
 
-  private native void createChannel(String address);
-  private native String sayHello(String name);
+  /** Calls the sayHello gRPC endpoint and returns the reply. */
+  private String sayHello(String address, String name) {
+    try {
+      return HelloWorldGrpc.newBlockingStub(createChannel(address))
+          .sayHello(HelloRequest.newBuilder().setGreeting(name).build())
+          .getReply();
+    } catch (Exception exception) {
+      Log.e("Oak", "Exception", exception);
+      return "Unexpected error!";
+    }
+  }
+
+  /** Creates a TLS channel. */
+  private ManagedChannel createChannel(String address) throws Exception {
+    return OkHttpChannelBuilder.forTarget(address)
+        .useTransportSecurity()
+        .sslSocketFactory(getSocketFactory())
+        .build();
+  }
 
-  private String rpcAddress;
+  /** Gets a socket factory configured with a custom CA certificate to trust. */
+  private SSLSocketFactory getSocketFactory() throws Exception {
+    InputStream rawCertificate = getAssets().open("ca.pem");
+    Certificate certificate =
+        CertificateFactory.getInstance("X.509").generateCertificate(rawCertificate);
+    KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    // No stored keystore file or password.
+    keyStore.load(null, null);
+    keyStore.setCertificateEntry("local_ca", certificate);
+    TrustManagerFactory trustManagerFactory =
+        TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+    trustManagerFactory.init(keyStore);
+    SSLContext tlsContext = SSLContext.getInstance("TLSv1.2");
```

Comment:
Aren't we using `TLSv1.3`?

Reply:
TLSv1.3 is only supported from API level 29. TLSv1.2 works with the current c++ gRPC pseudo node. I assume it is backward-compatible.

Reply:
Could you please add this comment to the code?

Reply:
Done.